File: Python/afe/algorithms/sorting/bubble_sort.py
# -*- coding: utf-8 -*-

from typing import List, TypeVar, Optional, Callable
from .sort_interface import SortInterface


T = TypeVar('T')


class BubbleSort(SortInterface):

    """
    A class that implements the Bubble Sort algorithm. \n
    Bubble Sort is a simple sorting algorithm that repeatedly steps through the list, compares adjacent elements,
    and swaps them if they are in the wrong order. The pass through the list is repeated until no swaps are needed,
    which means the list is sorted. \n
    It is a stable sort, meaning that it maintains the relative order of records with equal keys.
    """

    @staticmethod
    def sort(data: List[T], comparator: Optional[Callable[[T, T], int]] = None) -> None:
        
        """
        Sorts a list in-place using Bubble Sort. Stable.
        - Time Complexity: O(n^2)
        - Space Complexity: O(1)
        """
        
        if comparator is None: comparator = BubbleSort.DEFAULT_COMPARATOR

        n = len(data)
        for i in range(n):
            swapped = False
            for j in range(0, n - i - 1):
                if comparator(data[j], data[j + 1]) > 0:
                    data[j], data[j + 1] = data[j + 1], data[j]
                    swapped = True
            if not swapped:
                break

File: Java/src/main/java/io/github/mihaistreames/afe/algorithms/sorting/BubbleSort.java
package io.github.mihaistreames.afe.algorithms.sorting;

import org.jetbrains.annotations.NotNull;

import java.util.Comparator;
import java.util.List;

/**
 * A simple sorting algorithm that repeatedly steps through the list,
 * compares adjacent elements and swaps them if they are in the wrong order.
 * <p>
 * This implementation is not stable. <br>
 * Time Complexity: O(n^2) <br>
 * Space Complexity: O(1)
 */
public final class BubbleSort {

    private BubbleSort() {
        throw new UnsupportedOperationException("Utility class cannot be instantiated");
    }

    /**
     * Sorts a list using the bubble sort algorithm and natural ordering.
     *
     * @param <T>  the type of elements, must extend {@link Comparable}
     * @param list the list to be sorted
     */
    public static <T extends Comparable<T>> void bubbleSort(@NotNull final List<T> list) {
        sort(list, Comparator.naturalOrder());
    }

    /**
     * Sorts a list using the bubble sort algorithm and a custom comparator.
     *
     * @param <T>        the type of elements
     * @param list       the list to be sorted
     * @param comparator the comparator to determine the order
     */
    public static <T> void bubbleSort(@NotNull final List<T> list,
                                      @NotNull final Comparator<T> comparator) {
        sort(list, comparator);
    }

    private static <T> void sort(@NotNull final List<T> list,
                                 @NotNull final Comparator<T> comparator) {
        final int size = list.size();
        for (int i = 0; i < size - 1; i++) {
            for (int j = 0; j < size - i - 1; j++) {
                if (comparator.compare(list.get(j), list.get(j + 1)) > 0) {
                    final T temp = list.get(j);
                    list.set(j, list.get(j + 1));
                    list.set(j + 1, temp);
                }
            }
        }
    }
}